### PR TITLE
[codex] Make required seed data idempotent

### LINKED ISSUE #73: Make required reference seed data idempotent
Parent: #72

## Context

`app_service_data.sql`, `apps_data.sql`, and `app_metadata_data.sql` are required reference data for meaningful local ETL runs. Today they are loaded by bootstrap after schema reset, but the seed SQL itself is not independently safe to re-run against an existing schema.

## Goal

Make required seed loading explicit and re-execution safe so future bootstrap and migration flows can rely on deterministic reference data state.

## Tasks

- Define stable conflict keys for `app_service`, `apps`, and `app_metadata`.
- Convert required seed inserts to an idempotent strategy such as `ON CONFLICT DO UPDATE` or an equivalent safe load path.
- Keep required reference data separate from future dev/test sample data.
- Add tests or verification that seed loading can be repeated without duplicate rows or failures.

## Acceptance criteria

- Required seed data can be applied more than once without PK/unique conflicts.
- Expected counts remain stable after repeated seed execution.
- The local bootstrap flow can rely on the same seed files or seed entrypoint without hidden ordering mistakes beyond the documented dependency order.

### PR DESCRIPTION
**Pull request by hoonly01**
## Summary
- Add idempotent upserts to required app_service, apps, and app_metadata seed files
- Add app_metadata natural conflict key on (app_id, valid_from) in schema and ORM model
- Cover seed conflict handlers and repeated seed execution counts in bootstrap tests

## Validation
- PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py
- Result: 9 passed, 1 skipped, 1 warning
- The skipped test is the live PostgreSQL repeat-execution check because local test Postgres/Docker was unavailable in this environment.

Closes #73

## Summary by CodeRabbit

* **Improvements**
  * Seeded data is now conflict-aware: duplicate inserts are updated instead of failing, making bootstrapping idempotent.
  * A uniqueness rule prevents duplicate historical metadata entries for the same app and timestamp.

* **Tests**
  * New tests validate seed files can be run repeatedly and assert expected post-seed row counts.

### CODE REVIEW

File: sql/schema_v4.sql [hunk 1]
```diff
@@ -98,7 +98,8 @@ CREATE TABLE app_metadata
   valid_from DATE,
   valid_to   DATE,
   is_active  BOOLEAN,
-  PRIMARY KEY (id)
+  PRIMARY KEY (id),
+  UNIQUE (app_id, valid_from)
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply sql/schema_v4.sql]: Decision: Optional, implemented.

Rationale: The ORM already names this constraint as `uq_app_metadata_app_id_valid_from`; naming the raw SQL constraint the same way avoids schema/model drift and is useful for future migration work, without changing runtime seed behavior.

Commit: 47ecf6c

Change summary: Updated `sql/schema_v4.sql` from an unnamed `UNIQUE (app_id, valid_from)` clause to `CONSTRAINT uq_app_metadata_app_id_valid_from UNIQUE (app_id, valid_from)`.

Validation: `PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py` -> 9 passed, 1 skipped, 1 warning. The skipped test is the live PostgreSQL repeat-execution check because local test Postgres/Docker is unavailable in this environment.

File: sql/schema_v4.sql [hunk 2]
```diff
@@ -98,7 +98,8 @@ CREATE TABLE app_metadata
...
 );
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply sql/schema_v4.sql:103]: Decision: Optional, not implemented.

Rationale: The NULL behavior is real in PostgreSQL, but the required reference seed data always supplies `valid_from = 2025-01-01`, so this does not affect #73 seed re-execution safety. Making `valid_from` `NOT NULL` is a stronger SCD2 schema decision and was not part of the approved follow-up scope.

Commit: Not implemented.

Change summary: No code change for this thread.

Validation: The approved constraint-name change was validated with `PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py` -> 9 passed, 1 skipped, 1 warning.

File: tests/test_bootstrap_db.py [hunk 3]
```diff
@@ -16,6 +19,47 @@
...
+    return "\n".join(filtered_lines)
+
+
+def _reset_schema(test_db_engine) -> None:
+    raw_conn = test_db_engine.raw_connection()
+    try:
+        cursor = raw_conn.cursor()
+        cursor.execute("DROP SCHEMA IF EXISTS public CASCADE;")
+        cursor.execute("CREATE SCHEMA public;")
+        cursor.execute(_schema_sql_for_test())
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply tests/test_bootstrap_db.py:47]: Decision: Optional, not implemented.

Rationale: Extracting the duplicated schema filtering/reset logic would reduce future drift risk, but it is a test refactor and not required for #73 core behavior. The approved scope for this follow-up was limited to naming the SQL constraint consistently with the ORM.

Commit: Not implemented.

Change summary: No code change for this thread.

Validation: The approved constraint-name change was validated with `PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py` -> 9 passed, 1 skipped, 1 warning.

File: tests/test_bootstrap_db.py [hunk 4]
```diff
@@ -59,6 +103,46 @@ def test_build_verification_queries_check_required_seed_counts():
...
+                "app_metadata": conn.execute(
+                    text("SELECT COUNT(*) FROM app_metadata WHERE is_active = TRUE")
+                ).scalar_one(),
+            }
+
+        assert counts == {
+            "app_service": 39,
+            "apps": 63,
+            "app_metadata": 63,
+        }
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply tests/test_bootstrap_db.py:141]: Decision: Optional, not implemented.

Rationale: Reusing `build_verification_queries()` would reduce duplicated expected counts in tests, but the current hardcoded assertions match the existing verification contract and do not affect seed idempotency. This is a maintainability improvement outside the approved follow-up scope.

Commit: Not implemented.

Change summary: No code change for this thread.

Validation: The approved constraint-name change was validated with `PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py` -> 9 passed, 1 skipped, 1 warning.

File: tests/test_bootstrap_db.py [hunk 5]
```diff
@@ -59,6 +103,46 @@ def test_build_verification_queries_check_required_seed_counts():
     ]
 
 
+def test_required_seed_files_define_idempotent_conflict_handlers():
+    assert "ON CONFLICT (service_id) DO UPDATE" in (ROOT / "sql" / "app_service_data.sql").read_text()
+    assert "ON CONFLICT (app_id) DO UPDATE" in (ROOT / "sql" / "apps_data.sql").read_text()
+    assert "UNIQUE (app_id, valid_from)" in (ROOT / "sql" / "schema_v4.sql").read_text()
+    assert "ON CONFLICT (app_id, valid_from) DO UPDATE" in (
+        ROOT / "sql" / "app_metadata_data.sql"
+    ).read_text()
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply tests/test_bootstrap_db.py:112]: Decision: Optional, not implemented.

Rationale: The substring assertions are formatting-sensitive, but they are only a lightweight contract check. The behavioral coverage is the repeat-execution test, which applies the seed files against PostgreSQL when the test DB is available. Reworking this assertion style was not part of the approved follow-up scope.

Commit: Not implemented.

Change summary: No code change for this thread.

Validation: The approved constraint-name change was validated with `PYTHONPATH=. uv run pytest tests/test_bootstrap_db.py` -> 9 passed, 1 skipped, 1 warning.
